src/sync.cpp

```diff
@@ -135,16 +139,53 @@ static void potential_deadlock_detected(const LockPair& mismatch, const LockStac
     throw std::logic_error(strprintf("potential deadlock detected: %s -> %s -> %s", mutex_b, mutex_a, mutex_b));
 }
 
-static void push_lock(void* c, const CLockLocation& locklocation)
+static void double_lock_detected(const void* mutex, LockStack& lock_stack)
 {
+    LogPrintf("DOUBLE LOCK DETECTED\n");
+    LogPrintf("Lock order:\n");
+    for (const LockStackItem& i : lock_stack) {
+        if (i.first == mutex) {
+            LogPrintf(" (*)"); /* Continued */
+        }
+        LogPrintf(" %s\n", i.second.ToString());
+    }
+    // We throw or abort() here, never actually locking the mutex. So remove the entry
+    // from the lock stack which we just added in push_lock(). Tests in sync_tests.cpp
+    // continue executing past the exception and we don't want one test to leave a
+    // bricked stack for a subsequent test that could execute in the same thread.
+    lock_stack.pop_back();
+    if (g_debug_lockorder_abort) {
+        tfm::format(std::cerr, "Assertion failed: detected double lock at %s:%i, details in debug log.\n", __FILE__, __LINE__);
+        abort();
```

Comment:
```suggestion
        std::abort();
```

Should be the same symbol, but slightly clearer.

Reply:
There's more cases of s/abort()/std::abort so a follow up to to do this is also fine.

Reply:
I am leaving it as is for consistency with `potential_deadlock_detected()` which uses `abort()`. As @promag says, a subsequent patch could change all occurrences.